codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/customizations/RulesBasedAuthSchemeResolverGeneratorTests.kt

```diff
@@ -0,0 +1,151 @@
+                            validAuthOptions.append(AuthOption(schemeID: "smithy.api#httpBearerAuth"))
+                        case "onlyHttpApiKeyAndBearerAuthReversed":
+                            validAuthOptions.append(AuthOption(schemeID: "smithy.api#httpBearerAuth"))
+                            validAuthOptions.append(AuthOption(schemeID: "smithy.api#httpApiKeyAuth"))
+                        case "onlySigv4Auth":
+                            var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: "weather")
+                            guard let region = serviceParams.region else {
+                                throw ClientError.authError("Missing region in auth scheme parameters for SigV4 auth scheme.")
+                            }
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)
+                            validAuthOptions.append(sigV4Option)
+                        case "onlySigv4AuthOptional":
+                            var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: "weather")
+                            guard let region = serviceParams.region else {
+                                throw ClientError.authError("Missing region in auth scheme parameters for SigV4 auth scheme.")
+                            }
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)
+                            validAuthOptions.append(sigV4Option)
+                            validAuthOptions.append(AuthOption(schemeID: "smithy.api#noAuth"))
+                        case "onlyCustomAuth":
+                            validAuthOptions.append(AuthOption(schemeID: "com.test#customAuth"))
+                        case "onlyCustomAuthOptional":
+                            validAuthOptions.append(AuthOption(schemeID: "com.test#customAuth"))
+                            validAuthOptions.append(AuthOption(schemeID: "smithy.api#noAuth"))
+                        default:
+                            var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: "weather")
+                            guard let region = serviceParams.region else {
+                                throw ClientError.authError("Missing region in auth scheme parameters for SigV4 auth scheme.")
+                            }
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)
+                            validAuthOptions.append(sigV4Option)
+                    }
+                    return validAuthOptions
```

Comment:
You can combine regular and Optional case statements into one in order to reduce code repetition. For example let's take "onlyHttpApiKeyAuth" and "onlyHttpApiKeyAuthOptional" which both add httpApiKeyAuth you can do the following:
```
case "onlyHttpApiKeyAuth", "onlyHttpApiKeyAuthOptional":
    validAuthOptions.append(AuthOption(schemeID: Constants.httpApiKeyAuth))
    if serviceParams.operation.hasSuffix("Optional") { addNoAuth }
```

Here you can add the code snippet for addNoAuth or optionally you could consider making it a small function like so: `let addNoAuth = { validAuthOptions.append(AuthOption(schemeID: Constants.NoAuth)) }` and call `addNoAuth()`